File: mobile/test/core/protocol/secure_envelope_test.dart
import 'dart:typed_data';

import 'package:meshsetu_mobile/core/model/model.dart';
import 'package:meshsetu_mobile/core/protocol/secure_envelope.dart';
import 'package:test/test.dart';

void main() {
  test('authenticated round trip and tamper rejection', () async {
    final crypto = AeadEnvelope(List.generate(32, (i) => i));
    final source = MeshEnvelope(
      objectId: 8,
      eventId: 'event',
      siteId: 'site',
      roomId: 'room',
      createdAtMs: 1,
      expiresAtMs: 2,
      hopCount: 0,
      hopLimit: 5,
      priority: PriorityBand.p0Critical,
      payloadType: PayloadType.structuredSos,
      payload: Uint8List.fromList([1]),
      originEphemeralId: 9,
    );
    final encrypted = await crypto.encrypt(source);
    final decoded = await crypto.decrypt(encrypted);
    expect(decoded, isNotNull);
    expect(decoded!.objectId, source.objectId);
    expect(decoded.eventId, source.eventId);
    expect(decoded.payload, orderedEquals(source.payload));

    final tamperedBytes = Uint8List.fromList(encrypted.bytes);
    tamperedBytes[tamperedBytes.length - 1] ^= 1;
    final tampered = EncryptedObject(
      objectId: encrypted.objectId,
      trafficClass: encrypted.trafficClass,
      bytes: tamperedBytes,
      expiresAtMs: encrypted.expiresAtMs,
    );
    expect(await crypto.decrypt(tampered), isNull);
  });

  test('site metadata is authenticated', () async {
    final crypto = AeadEnvelope(List.generate(32, (i) => i));
    final source = MeshEnvelope(
      objectId: 9,
      eventId: 'event',
      siteId: 'site',
      roomId: 'room',
      createdAtMs: 1,
      expiresAtMs: 2,
      hopCount: 0,
      hopLimit: 5,
      priority: PriorityBand.p0Critical,
      payloadType: PayloadType.structuredSos,
      payload: Uint8List.fromList([1]),
      originEphemeralId: 9,
    );
    final encrypted = await crypto.encrypt(source);
    final tamperedBytes = Uint8List.fromList(encrypted.bytes);
    // Byte 3 is the first byte of the site-id region (after 1-byte version +
    // 2-byte length prefix); it's authenticated via AAD but not itself
    // encrypted, so tampering it must still fail decryption.
    tamperedBytes[3] ^= 0xFF;
    final tampered = EncryptedObject(
      objectId: encrypted.objectId,
      trafficClass: encrypted.trafficClass,
      bytes: tamperedBytes,
      expiresAtMs: encrypted.expiresAtMs,
    );
    expect(await crypto.decrypt(tampered), isNull);
  });
}

File: core-protocol/src/test/kotlin/in/meshsetu/protocol/SecureEnvelopeTest.kt
package `in`.meshsetu.protocol

import `in`.meshsetu.model.MeshEnvelope
import `in`.meshsetu.model.PayloadType
import `in`.meshsetu.model.PriorityBand
import java.util.Base64
import javax.crypto.spec.SecretKeySpec
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import kotlin.test.assertContentEquals

class SecureEnvelopeTest {
    private val crypto = AeadEnvelope(SecretKeySpec(ByteArray(32) { it.toByte() }, "AES"))

    @Test fun authenticatedRoundTripAndTamperRejection() {
        val source = MeshEnvelope(8u, "event", "site", "room", 1, 2, 0, 5, PriorityBand.P0_CRITICAL, PayloadType.STRUCTURED_SOS, byteArrayOf(1), 9u)
        val encrypted = crypto.encrypt(source)
        val decoded = crypto.decrypt(encrypted).getOrThrow()
        assertEquals(source.objectId, decoded.objectId)
        assertEquals(source.eventId, decoded.eventId)
        assertContentEquals(source.payload, decoded.payload)
        val tampered = encrypted.copy(bytes = encrypted.bytes.copyOf().also { it[it.lastIndex] = (it.last().toInt() xor 1).toByte() })
        assertTrue(crypto.decrypt(tampered).isFailure)
    }

    @Test fun siteMetadataIsAuthenticated() {
        val source = MeshEnvelope(9u, "event", "site", "room", 1, 2, 0, 5, PriorityBand.P0_CRITICAL, PayloadType.STRUCTURED_SOS, byteArrayOf(1), 9u)
        val encrypted = crypto.encrypt(source)
        val tampered = encrypted.copy(bytes = encrypted.bytes.copyOf().also { it[3] = 'x'.code.toByte() })
        assertTrue(crypto.decrypt(tampered).isFailure)
    }
}
